T2Ramsey tutorial

merav-aharoni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Tutorial for T2 Ramsey experiment.


### Details and comments

merav-aharoni commented:
Ready for review.

yaelbh commented on docs/tutorials/backends/t2star_backend.py:
```diff
@@ -0,0 +1,128 @@
+import numpy as np
```
`t2star_beckend.py` should be located with the other mock backends in `qiskit_experiments/test`. It's temporary anyway because it will be replaced by Ethan's backend.

merav-aharoni replied:
Done.

yaelbh commented on docs/tutorials/backends/t2star_backend.py:
```diff
@@ -0,0 +1,128 @@
+
+
+# Fix seed for simulations
+SEED = 9000
```
Is the seed to be fixed here or in the test file?

merav-aharoni replied:
I think it should be fixed in both.

yaelbh commented on docs/tutorials/backends/t2star_backend.py:
```diff
@@ -0,0 +1,128 @@
+                    "data": {"counts": counts},
+                }
+            )
+        return Result.from_dict(result)
```
In the main branch I think it was replaced by a `MockJob` (to be probably replaced again by `FakeJob` by #115, to be replaced by Ethan's).

merav-aharoni replied:
Yes, I just realized I didn't update this in the tutorial.

merav-aharoni replied:
Done, along with several other updates that were missing.

yaelbh commented on docs/tutorials/t2Ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,415 @@
+{
```
Is it intentional that there are two notebooks? Possibly `t2ramsey_tutorial` with small `r` is better.

merav-aharoni replied:
Strange, I don't see a `t2Ramsey_tutorial.ipynb` with capital `R`. I did find a `t2star_tutorial.ipynb` and removed it.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This experiment serves as one of the series of experiments used to characterize a single qubit. Its purpose is to determine two of the qubit's properties: *detuning frequency* and *$T_2*$*. The rough frequency of the qubit was already determined previously. Here, we would like to measure the *detuning*, that is  the difference between the qubit's precise frequency and the frequency of the ro ...
```
This cell is not being rendered correctly - starting at the first *$T_2*$*

merav-aharoni replied:
I don't see a problem here when I open it. @yaelbh , can you try?

yaelbh replied:
I didn't check on our machine, but here in GitHub it looks like this:
![image](https://user-images.githubusercontent.com/22316088/123808021-0a86e100-d8f9-11eb-8731-cecc672c8ceb.png)

eliarbel commented:
Please mark as WIP until it's ready for final review

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "metadata": {},
+   "source": [
+    "The resulting graph will have the form:\n",
+    "$ f(t) = a^{-x/T_2*} \\cdot cos(2 \\pi f t + \\phi) + b $\n",
```
replace `x` with `t`

merav-aharoni replied:
Right! changed.

merav-aharoni commented:
I thought it was ready for review, but when I looked now, I see one of the examples disappeared. Very mysterious. I will put back the WIP until fixed.

merav-aharoni commented:
I think this can be reviewed now. I am still not sure what to do with the last plot. @eliarbel - can you review all, in particular, do you think any more examples should be added?

yaelbh commented:
Fix the font size in the title:
![image](https://user-images.githubusercontent.com/22316088/124113011-9ae93100-da73-11eb-8b60-8ca87e59bd49.png)

yaelbh commented:
> when the qubits is initiated

Change "qubits" to singular. Maybe "initialized" instead of "initiated"?

yaelbh commented:
It is better to use reasonable numbers from the real world. Since T2 is typically in the area of ~100us, it doesn't make sense to take delays from 1s to 50s.

yaelbh commented:
... and to simulate T2 of 10s

yaelbh commented:
```
result2 = expdata2.analysis_result(0)
```
Same comment as in the "add tests" PR - this is actually the original result, you need `analysis_result(1)` for the combined result

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# T_2 Ramsey Experiment"
```
can write `# $T_2$ Ramsey Experiment`, math will work here as well

eliarbel replied:
Also I'm a bit confused about whether we should call it $T_2$ or $T_2*$

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+    "\n",
+    "    1. Hadamard gate\n",
+    "    2. delay\n",
+    "    3. p gate that rotates the qubit in the x-y plane \n",
```
Change to `P (phase) gate... "

merav-aharoni replied:
done.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+    "    4. Hadamard gate\n",
+    "    5. measurement\n",
+    "\n",
+    "During the delay time, we expect the qubit to precess about the z-axis. If the p gate and the precession offset each other perfectly, then the qubit will arrive at the |1⟩ state (after the second Hadamard gate). By varying the extension of the delays, we get a series of oscillations of the qubit state between the |0⟩ and |1⟩ states. We can draw the graph of the resulting function, and can an ...
```
> then the qubit will arrive at the |1⟩ state (after the second Hadamard gate)

Since you are not doing X90-deley-X90 but rather H-delay-H, the qubit returns to |0> (with perfect tuning), isn't it ?

merav-aharoni replied:
You are right - nice catch!

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We run the experiment on a simple, simulated backend, created specifically for this experiment."
```
Can be even more specific and say "created specifically for this experiment's tutorial"

merav-aharoni replied:
okay.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+    "The resulting graph will have the form:\n",
+    "$ f(t) = a^{-t/T_2*} \\cdot cos(2 \\pi f t + \\phi) + b $\n",
+    "where *t* is the delay, $T_2*$ is the decay factor, and *f* is the detuning frequency.\n",
+    "`conversion_factor` is a scaling factor that depends on the STI units used. It is 1E-6 here, because the unit is microseconds."
```
Explain (or just expand the acronym) what is STI

merav-aharoni replied:
ok, done.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+    }
+   ],
+   "source": [
+    "from qiskit_experiments.characterization import T2RamseyAnalysis\n",
```
Add  sub title for this section, something like "Providing initial user estimates". Also would be nice to give 1-2 sentences about where initial estimates can come from

merav-aharoni replied:
Agree. Done.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The units can be changed, but the output in the result is always given in seconds. The units in the backend must be adjusted accordingly."
```
> The units in the backend must be adjusted accordingly.

This is just for the mock backend right ? Need to make explicit since real backends are different

yaelbh replied:
I'd drop this line, considering that it's not a tutorial about how to use the mock backend

eliarbel replied:
Pedagogically, I think the mock backend is useful, when one wants a quick learning TATs. I think even a dedicated sub-section for the mock backend is in place (documenting how to tweak it, nothing beyond what is already in the tutorial). This way there is a clean separation between how to use the experiment, and some dedicated reference for the mock backend, allowing users to focus on what they need easily

yaelbh replied:
This makes sense. However if the backend changes to Ethan's soon then it may not be worth the effort.

merav-aharoni replied:
This is all probably temporary, so not worth investing too much effort here. The reason I mention the mockjob is because I must import it, and I don't like writing a line that cannot be understood as is.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It is possible to add data to an experiment, after the analysis of the first set of data. In the next example we add exp2 to `exp_in_ns` that we showed above."
```
Add subtitle, something like "Adding data to an existing experiment"

merav-aharoni replied:
ok, done.

eliarbel commented on docs/tutorials/t2ramsey_tutorial.ipynb:
```diff
@@ -0,0 +1,416 @@
+     ]
+    },
+    {
+     "data": {
```
Right, the connecting line is a bummer here... no luck I guess removing it so far. :-(

yaelbh replied:
The connecting line is not supposed to be here anymore, since the T1 tutorial PR has been merged

yaelbh replied:
What's the blue line?

merav-aharoni replied:
Indeed the redundant lines disappeared. What did you do in T1 that fixed it?

yaelbh replied:
I removed all the lines, it's only scatter now

eliarbel reviewed (changes requested):
Looks very nice. I think a few final touches and it's good to go

yaelbh commented:
Fix this:
```
# set the computation units to seconds
unit = 'us' #seconds
```

yaelbh commented:
https://github.com/Qiskit/qiskit-experiments/pull/122#issuecomment-872144988 is still relevant

merav-aharoni commented:
@yaelbh - for some reason I don't have response boxes beneath your comments, so I am responding here:
1. I fixed all the units to be realistic
2. Changed the comment from `seconds` to `microseconds`
3. Fixed the result in the last example to be result(1)

yaelbh commented:
https://github.com/Qiskit/qiskit-experiments/pull/122#issuecomment-873598220
Text says "seconds" in two places instead of microseconds

merav-aharoni commented:
@yaelbh , @eliarbel - thank you both for all your thoughtful comments. I think I addressed them all. One more thing I need to do, is to improve the scaling of the x-axes of the graphs. I'll take care of that before a final review.

merav-aharoni commented:
@yaelbh , @eliarbel - I addressed all your comments. It is ready for one final review in my opinion.

yaelbh commented:
@eliarbel Your approval is required, because you asked for changes

eliarbel commented:
> @eliarbel Your approval is required, because you asked for changes

Just need to change T_2 to $T_2$ in the first header. I'll approve anyway, but please see that it's changed since not sure when Merav can handle this.
